### PR TITLE
[hma][api] AdditionalMatchSettingsConfig adjustable from API and match distance in /for-hash/ results

### PR DESCRIPTION
**Pull request by BarrettOlson**
Summary
---------

Depends on #885 and #883 (and has their commits)

Take the distance recorded in #883 and also adds it to `/for-hash/`

Provides endpoints on the `/datasets/` path for 
- listing all     
  - GET `/datasets/match-settings`
- listing one
  - GET `/datasets/match-settings/<key>`
- creating/updating 
  - POST `/datasets/match-settings` `{  "privacy_group_id": <key>, "pdq_match_threshold": <val>}`
- deleting
  - DELETE `/datasets/match-settings/<key>`

Test Plan
---------
Tested all new dataset endpoints and new response format for `/matches/for-hash/`

`https://{{api_id}}.execute-api.us-east-1.amazonaws.com/api/matches/for-hash/?signal_type=pdq&signal_value=f8f8f0cee0f4a84f06370a22038f63f0b36e2ed596621e1d33e6b39c4e9c9b22`
```json
{
    "matches": [
        {
            "match_distance": 0,
            "matched_signal_fields": {
                "signal_id": "4109214869142910",
                "privacy_group_id": "inria-holidays-test",
                "signal_type": "pdq",
                "signal_hash": "f8f8f0cee0f4a84f06370a22038f63f0b36e2ed596621e1d33e6b39c4e9c9b22",
                "tags": [
                    "test_pdq_samples",
                    "holidays_jpg1_dataset",
                    "but_not_really_it_is_an_extra_photo_see_b_jpg",
                    "disputed"
                ]
            }
        }
    ]
}
```

### CODE REVIEW

File: hasher-matcher-actioner/hmalib/lambdas/api/datasets.py [hunk 1]
```diff
@@ -294,4 +339,66 @@ def delete_dataset(key=None) -> DeleteDatasetResponse:
...
+    ) -> MatchSettingsUpdateResponse:
+        """
+        Create or update a match settings config for a given privacy_group_id
+        """
+        if config := AdditionalMatchSettingsConfig.get(request.privacy_group_id):
+            config.pdq_match_threshold = request.pdq_match_threshold
+            hmaconfig.update_config(config)
+
+            return MatchSettingsUpdateResponse(
+                f"match_settings updated for pg_id {request.privacy_group_id} with pdq_match_threshold={request.pdq_match_threshold}"
```
Comment (schatten) [inline hasher-matcher-actioner/hmalib/lambdas/api/datasets.py:379]: nit: Personally, don't see the value of logging style messages in responses. A 200 is enough if you don't have anything to return.

File: hasher-matcher-actioner/hmalib/lambdas/api/datasets.py [hunk 2]
```diff
@@ -294,4 +339,66 @@ def delete_dataset(key=None) -> DeleteDatasetResponse:
...
+            )
+
+        config = AdditionalMatchSettingsConfig(
+            request.privacy_group_id, request.pdq_match_threshold
+        )
+        hmaconfig.create_config(config)
+        return MatchSettingsUpdateResponse(
+            f"match_settings created for pg_id {request.privacy_group_id} with pdq_match_threshold={request.pdq_match_threshold}"
```
Comment (schatten) [inline hasher-matcher-actioner/hmalib/lambdas/api/datasets.py:387]: ditto.

File: hasher-matcher-actioner/hmalib/lambdas/api/datasets.py (hunk 1 above)
Comment (schatten) [inline hasher-matcher-actioner/hmalib/lambdas/api/datasets.py:379]: nit: Personally, don't see the value of logging style messages in responses. A 200 is enough if you don't have anything to return.

File: hasher-matcher-actioner/hmalib/lambdas/api/matches.py [hunk 3]
```diff
@@ -196,22 +196,22 @@ def from_dict(cls, d):
 
 
 @dataclass
-class MatchesForHashResponse(JSONifiable):
-    matches: t.List[ThreatExchangeSignalMetadata]
-    signal_value: str
+class MatchesForHash(JSONifiable):
+    match_distance: int
+    matched_signal_fields: ThreatExchangeSignalMetadata  # or matches signal from other sources
```
Comment (schatten) [inline hasher-matcher-actioner/hmalib/lambdas/api/matches.py:201]: struggling to understand the name here. Why are these matched signal fields?

`matched_signals` instead?

**Review by schatten — APPROVED** [review]
Q about `matched_signal_fields`. What does that mean?

personal nit about responses containing log messages.

File: hasher-matcher-actioner/hmalib/lambdas/api/datasets.py (hunk 1 above)
Comment (BarrettOlson) [reply hasher-matcher-actioner/hmalib/lambdas/api/datasets.py:379]: Agree! but until we have UI support for changes I find it helpful and worth keeping for now. (you deleted the comment haha)

File: hasher-matcher-actioner/hmalib/lambdas/api/matches.py (hunk 3 above)
Comment (BarrettOlson) [reply hasher-matcher-actioner/hmalib/lambdas/api/matches.py:201]: fair!

Comment (BarrettOlson) [reply hasher-matcher-actioner/hmalib/lambdas/api/matches.py:201]: `matched_signal` because it is a single matched signal

**Pull request merged**